mmtk/src/api.rs

```diff
@@ -290,6 +291,37 @@ pub extern "C" fn process(name: *const c_char, value: *const c_char) -> bool {
     )
 }
 
+// We trust the name/value pointer is valid.
+#[allow(clippy::not_unsafe_ptr_arg_deref)]
+fn set_hotspot_flag_impl<T: Display>(name: *const c_char, value: T) {
+    let name_cstr: &CStr = unsafe { CStr::from_ptr(name) };
+    let name = name_cstr.to_str().unwrap();
+    let mut builder = BUILDER.lock().unwrap();
+    let value_str = format!("{}", value);
+    let result = match name {
+        "ParallelGCThreads" => builder.set_option("threads", &value_str),
+        "UseTransparentHugePages" => builder.set_option("transparent_hugepages", &value_str),
+        _ => unimplemented!("Unsupported flag: {}", name),
+    };
+    assert!(
+        result,
+        "Failed to pass hotspot flag {}={} to mmtk-core",
+        name, value
+    );
+}
+
+/// Pass a hotspot integer command line flag to mmtk
+#[no_mangle]
+pub extern "C" fn mmtk_set_hotspot_flag_uint(name: *const c_char, value: u32) {
+    set_hotspot_flag_impl(name, value)
+}
+
+/// Pass a hotspot boolean command line flag to mmtk
+#[no_mangle]
+pub extern "C" fn mmtk_set_hotspot_flag_bool(name: *const c_char, value: bool) {
+    set_hotspot_flag_impl(name, value)
+}
+
```

Comment:
In fact the `options` field in `MMTKBuilder` is mutable, and can be set in a strongly typed way.  And the code is quite simple.  It is pointless to parse options by name.

```suggestion
#[no_mangle]
pub extern "C" fn mmtk_builder_set_threads(value: usize) {
    let mut builder = BUILDER.lock().unwrap();
    builder.options.threads.set(value);
}

#[no_mangle]
pub extern "C" fn mmtk_builder_set_transparent_hugepages(value: bool) {
    let mut builder = BUILDER.lock().unwrap();
    builder.options.transparent_hugepages.set(value);
}

```

Reply:
Thanks. Fixed. I didn't release there is a typed interface to do that.

Reply:
> ... I didn't release there is a typed interface to do that.

It was made possible since https://github.com/mmtk/mmtk-core/pull/625.  Before that, the `Options` struct was within an `UnsafeOptionsWrapper` which only allowed the string-based setters.  After that, the `Options` is mutable inside `MMTKBuilder`, and immutable in `MMTK`.  That allows it to be manipulated directly when still in `MMTKBuilder`.